### PR TITLE
CollisionParabola into CollisionBox

### PR DESCRIPTION
**Pull request by hecris**
Added collision test for parabola into box as well as unit tests. The algorithm iterates through the 6 faces of the box, testing their intersection with the parabola and finding the minimum 't' value aka the earliest point of intersection. Suggestions are welcomed!

### CODE REVIEW

**Comment by Moguri** [comment]
This looks good to me. @rdb, did you want to look over this?

**Comment by hecris** [comment]
Ok, got it thanks. Will make the changes soon.

**Comment by hecris** [comment]
I modified the algorithm to check if the intersection point is behind all of the other planes (by checking if distance to plane < 0). Also, if the parabola intersects from inside the box, then the surface normal should go inward (simply negate the vector). Like so:
It is currently passing all tests and when using @rdb 's visualizer it works for the most part.
The part where it fails is:
image
In this example, the parabola first intersects with face 5, but the distance from the point of intersection to the second plane (face 1) is positive, indicating that the point is in front of the second plane (which visually is not true):
image

If the mouse is moved just a little bit further, it intersects:
image

This part has been troubling me for quite a while and I can't seem to find the source of the problem. I also tried to use `CollisionBox::point_is_inside()` but the result was the same. Let me know if you have any suggestions.

File: panda/src/collide/collisionBox.cxx [hunk 1]
```diff
@@ -493,6 +494,102 @@ test_intersection_from_ray(const CollisionEntry &entry) const {
...
+        }
+        if (get_plane(j).dist_to_plane(local_p.calc_point(cur_t)) > 0.0f) {
+          // our point is in front of this face, so turns out
+          // it does not collide with the box at this point
+          behind = false;
+          break;
+        }
+      }
+
+      if (behind) {
```
Comment (Moguri) [inline panda/src/collide/collisionBox.cxx]: This probably does not solve your current problem, but shouldn't this check if `cur_t` is smaller than `t` so we get the first `intersecting_face` and `t` value?

File: panda/src/collide/collisionBox.cxx [hunk 2]
```diff
@@ -493,6 +494,102 @@ test_intersection_from_ray(const CollisionEntry &entry) const {
...
+    PT(CollisionEntry) new_entry = new CollisionEntry(entry);
+
+    LPlane face = get_plane(intersecting_face);
+
+    LPoint3 into_intersection_point = local_p.calc_point(t);
+    LVector3 normal = (has_effective_normal() && parabola->get_respect_effective_normal()) ? get_effective_normal() : face.get_normal();
+
+    // consider a point just before our intersection point
+    float eps = 1.0e-6f;
+    if (get_plane(intersecting_face).dist_to_plane(local_p.calc_point(t - eps)) < 0.0f) {
```
Comment (Moguri) [inline panda/src/collide/collisionBox.cxx]: `get_plane(intersecting_face)` is already stored in `face`

Comment (hecris) [reply panda/src/collide/collisionBox.cxx]: Thanks!

File: panda/src/collide/collisionBox.cxx (hunk 1 above)
Comment (hecris) [reply panda/src/collide/collisionBox.cxx]: The check is actually done beforehand on lines 520 and 525:
https://github.com/panda3d/panda3d/pull/664/files#diff-6105c95465dfdba16cebccbd91fdd51eR520

Comment (Moguri) [reply panda/src/collide/collisionBox.cxx]: Ah up, I see it now.

**Comment by Moguri** [comment]
@rdb Any further thoughts on the issue with the corner? If there isn't something obvious, I wonder if this is "good enough." Ideally our collision tests would be "perfect," but I know that is already not the case (e.g., I seem to remember you saying something about some of the capsule tests).

**Comment by hecris** [comment]
@Moguri @rdb It seems like the algorithm is working correctly. Although the parabola is appearing to collide with the box, it actually does not. Here's a visualizer:
image

- As you can see, the parabola first intersects with the vertical plane, despite appearing to intersect with the horizontal plane first. (you can verify this by looking at the surface normals). 
- When the parabola intersects with the vertical plane, the point of intersection is not behind the horizontal plane. (you can verify this by looking at the box's face)
- Similarly, when the parabola intersects with the horizontal plane, the point of intersection is not behind the vertical plane. Overall, it returns no collision.
I'll go ahead and add some more unit tests to verify this. In the meantime, let me know what you guys think. Also, I think I will move onto parabola into inverse sphere soon.

**Comment by rdb** [comment]
Could this be a perspective issue?  Does setting the lens to an `OrthographicLens()` fix it?

**Comment by rdb** [comment]
Also, I don't think the collision normal should point inwards.  It should always point outward, I think.

**Comment by hecris** [comment]
The collision test seems to work perfectly using `OrthographicLens()` !
And OK, I will make the surface normal go outwards.

**Comment by rdb** [comment]
It's working perfectly, thanks so much!  Merged.

**Pull request closed without merging**